File: 2 - Work/Phase 1 Parsing/Bulk Tests/python/01 - parse_and_spatial_join.py
# Purpose: Parse raw USA golf course CSV via regex into structured fields,
#          spatially join to 2022 US county boundaries via point-in-polygon,
#          and save the enriched dataset.
# Inputs:  00 - Data Sources/Original Data/Golf Courses-USA.csv
# Outputs: Bulk Tests/python/Py_Phase1_Spatial_Joined_Golf_Courses.csv


# === 1. LIBRARIES ===

import re
from pathlib import Path

import geopandas as gpd
import pandas as pd
from pygris import counties


# === 2. GLOBALS & PATHS ===

SCRIPT_DIR = Path(__file__).parent
ROOT_DIR   = SCRIPT_DIR.parent.parent.parent

RAW_CSV = ROOT_DIR / "00 - Data Sources" / "Original Data" / "Golf Courses-USA.csv"
OUT_CSV = SCRIPT_DIR / "Py_Phase1_Spatial_Joined_Golf_Courses.csv"


# === 3. FUNCTIONS ===

def extract_course_type(detail_str: str) -> str:
    """Classify ownership from the first parenthetical keyword in the Details field."""
    s = str(detail_str).lower()
    for label in ("public", "private", "municipal", "military", "resort"):
        if label in s:
            return label.title()
    return "Unknown"


def extract_holes(detail_str: str) -> int:
    """Return hole count from the Details field; defaults to 18 when absent."""
    m = re.search(r"\((\d+)\s*Holes?\)", str(detail_str), re.IGNORECASE)
    return int(m.group(1)) if m else 18


# === 4. EXECUTION ===

def main():
    if not RAW_CSV.exists():
        raise FileNotFoundError(f"Input file not found: {RAW_CSV}")
    courses_df = pd.read_csv(RAW_CSV, header=None,
                             names=["Longitude", "Latitude", "Course_Name", "Details"])

    courses_df["Ownership_Type"] = courses_df["Details"].apply(extract_course_type)
    courses_df["Holes"]          = courses_df["Details"].apply(extract_holes)

    courses_df  = courses_df.dropna(subset=["Longitude", "Latitude"])
    courses_geo = gpd.GeoDataFrame(
        courses_df,
        geometry=gpd.points_from_xy(courses_df["Longitude"], courses_df["Latitude"]),
        crs="EPSG:4326",
    )

    county_geo = counties(cb=True, year=2022, resolution="20m")
    county_geo = county_geo.to_crs("EPSG:4326")  # [METHODOLOGY] align CRS to WGS 84 — golf course coordinates are in EPSG:4326

    courses_geo = gpd.sjoin(  # [METHODOLOGY]
        courses_geo,
        county_geo[["GEOID", "NAME", "STATE_NAME", "geometry"]],
        how="left",
        predicate="intersects",
    )
    courses_geo = courses_geo.rename(columns={
        "GEOID":      "FIPS",
        "NAME":       "County_Name",
        "STATE_NAME": "State_Abbr",
    })
    if "index_right" in courses_geo.columns:
        courses_geo.drop(columns=["index_right"], inplace=True)

    total   = len(courses_geo)
    missing = courses_geo["FIPS"].isna().sum()

    print("\n=== OUTPUT STATISTICS ===")
    print(f"  Total rows:                {total:,}")
    print(f"  Missing FIPS (no county):  {missing}")
    print(f"\n  First 5 rows:")
    print(courses_geo[["Course_Name", "FIPS", "County_Name", "State_Abbr",
                        "Ownership_Type", "Holes"]].head().to_string(index=False))

    courses_geo.drop(columns=["geometry"]).to_csv(OUT_CSV, index=False)
    print(f"\n  [OK] Saved -> {OUT_CSV}")


if __name__ == "__main__":
    main()

File: 2 - Work/Phase 1 Parsing/Bulk Tests/R/01 - parse_and_spatial_join.R
# Purpose: Parse raw USA golf course CSV, extract course type and hole count,
#          and spatially join to US county boundaries via point-in-polygon.
# Inputs:  00 - Data Sources/Original Data/Golf Courses-USA.csv
# Outputs: Bulk Tests/R/R_Phase1_Spatial_Joined_Golf_Courses.csv


# === 1. LIBRARIES ===

suppressPackageStartupMessages({
  library(tidyverse)
  library(wooldridge)
  library(sf)
  library(tigris)
  library(this.path)
})


# === 2. GLOBALS & PATHS ===

options(tigris_use_cache = TRUE)

SCRIPT_DIR <- this.path::this.dir()
ROOT_DIR   <- file.path(SCRIPT_DIR, "..", "..", "..")

RAW_CSV <- file.path(ROOT_DIR, "00 - Data Sources", "Original Data", "Golf Courses-USA.csv")
OUT_CSV <- file.path(SCRIPT_DIR, "R_Phase1_Spatial_Joined_Golf_Courses.csv")


# === 3. FUNCTIONS ===

#' Extract standardized course type from raw details string.
#' @param details character vector of raw course detail strings
#' @return character vector of course type labels
extract_course_type <- function(details) {
  s <- tolower(details)
  case_when(
    str_detect(s, "public")    ~ "Public",
    str_detect(s, "private")   ~ "Private",
    str_detect(s, "municipal") ~ "Municipal",
    str_detect(s, "military")  ~ "Military",
    str_detect(s, "resort")    ~ "Resort",
    TRUE                       ~ "Unknown"
  )
}

#' Extract hole count from raw details string; defaults to 18 when absent.
#' @param details character vector of raw course detail strings
#' @return integer vector of hole counts
extract_holes <- function(details) {
  m     <- str_match(details, "\\((\\d+)\\s*Holes?\\)")
  holes <- as.integer(m[, 2])
  ifelse(is.na(holes), 18L, holes)
}


# === 4. EXECUTION ===

if (!file.exists(RAW_CSV)) stop(paste("Input file not found:", RAW_CSV))

cat("1: Loading raw Golf Courses CSV\n")
courses_df <- read_csv(RAW_CSV,
  col_names = c(
    "Longitude", "Latitude",
    "Course_Name", "Details"
  ),
  show_col_types = FALSE
)
cat(sprintf("    Rows loaded: %s\n", formatC(nrow(courses_df), big.mark = ",")))

cat("2: Extracting Course_Type & Holes via regex\n")
courses_df <- courses_df |>
  mutate(
    Course_Type = extract_course_type(Details),
    Holes       = extract_holes(Details)
  )
cat("    Course_Type distribution:\n")
print(table(courses_df$Course_Type))

cat("3: Converting to sf object (EPSG:4326)\n")
courses_df <- courses_df |> filter(!is.na(Longitude), !is.na(Latitude))
courses_sf <- st_as_sf(courses_df,
  coords = c("Longitude", "Latitude"), crs = 4326,
  remove = FALSE
)

cat("4: Downloading 2022 US County boundaries (tigris)\n")
# tigris cb=TRUE provides NAME (county) and STUSPS (state abbreviation) — renamed for cross-language consistency with Python output
county_sf <- counties(cb = TRUE, year = 2022, resolution = "20m") |>
  st_transform(4326) |> # [METHODOLOGY]
  select(GEOID, NAME, STATE_NAME = STUSPS)

cat("5: Spatial point-in-polygon join\n")
courses_sf <- st_join(courses_sf, county_sf, join = st_intersects) # [METHODOLOGY]

courses_sf <- courses_sf |>
  rename(FIPS = GEOID, County_Name = NAME, State_Abbr = STATE_NAME)

# CSV format cannot carry geometry — join attributes are retained as columns
courses_df <- st_drop_geometry(courses_sf)

total   <- nrow(courses_df)
missing <- sum(is.na(courses_df$FIPS))

cat("\nOUTPUT STATISTICS\n")
cat(sprintf("  Total rows:                %s\n", formatC(total, big.mark = ",")))
cat(sprintf("  Missing FIPS (no county):  %d\n", missing))
cat("\n  First 5 rows:\n")
print(head(courses_df |> select(
  Course_Name, FIPS, County_Name,
  State_Abbr, Course_Type, Holes
), 5))

write_csv(courses_df, OUT_CSV)
cat(sprintf("\n  [OK] Saved -> %s\n", OUT_CSV))
